plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+
+        let fetchEventsError: Error | unknown | null = null
+        try {
+            // :TODO: Verify this handles borders correctly
+            events = await fetchEventsForInterval(
+                hub.db,
+                pluginConfig.team_id,
+                new Date(payload.timestampCursor),
+                payload.offset,
+                payload.fetchTimeInterval,
+                EVENTS_PER_RUN
+            )
+        } catch (error) {
+            fetchEventsError = error
+            Sentry.captureException(error)
+        }
+
+        let exportEventsError: Error | unknown | null = null
+
+        if (fetchEventsError) {
+            // :TODO: This doesn't stop currently running jobs from continuing
+            await meta.storage.del(EXPORT_RUNNING_KEY)
+            // :TODO: Retries logic.
+            createLog(`Failed fetching events. Stopping export - please try again later.`)
+            return
+        } else {
+            if (events.length > 0) {
+                try {
+                    await methods.exportEvents!(events)
+                } catch (error) {
+                    exportEventsError = error
+                }
+            }
+        }
+
+        if (exportEventsError instanceof RetryError) {
+            const nextRetrySeconds = 2 ** payload.retriesPerformedSoFar * 3
+
+            // "Failed processing events 0-100 from 2021-08-19T12:34:26.061Z to 2021-08-19T12:44:26.061Z. Retrying in 3s"
+            createLog(
+                `Failed processing events ${payload.offset}-${payload.offset + events.length} from ${new Date(
+                    payload.timestampCursor
+                ).toISOString()} to ${new Date(
+                    payload.timestampCursor + payload.fetchTimeInterval
+                ).toISOString()}. Retrying in ${nextRetrySeconds}s`
+            )
+
+            await meta.jobs
+                .exportHistoricalEvents({
+                    ...payload,
+                    retriesPerformedSoFar: payload.retriesPerformedSoFar + 1,
+                } as ExportHistoricalEventsJobPayload)
+                .runIn(nextRetrySeconds, 'seconds')
+        } else if (!exportEventsError) {
+            const incrementCursor = events.length < EVENTS_PER_RUN
+            const incrementedTimeCursor = Math.min(payload.endTime, payload.timestampCursor + payload.fetchTimeInterval)
+
+            await meta.jobs
+                .exportHistoricalEvents({
+                    ...payload,
+                    retriesPerformedSoFar: 0,
+                    timestampCursor: incrementCursor ? incrementedTimeCursor : payload.timestampCursor,
+                    intraIntervalOffset: incrementCursor ? 0 : payload.offset + EVENTS_PER_RUN,
+                    fetchTimeInterval: nextFetchTimeInterval(payload, events.length),
+                } as ExportHistoricalEventsJobPayload)
+                .runIn(1, 'seconds')
+        }
+        // :TODO: When unknown exportEventsError?
```

Comment:
I think maybe we err on the side of protecting the system here and try to qualify errors as much as possible as retriable errors or not (both in the plugin code and plugin server code), and then just halt everything if we encounter a non-retriable error?

Or maybe again a strikes system with batch skipping and appropriate feedback to the user?

Just documenting thoughts